include/kernels/gauge_wilson_flow.cuh

```diff
@@ -56,48 +56,65 @@ namespace quda
     }
   };
 
-  template <typename Arg>
-  __host__ __device__ inline auto computeStaple(const Arg &arg, const int *x, int parity, int dir)
+  template <typename Arg> struct computeStapleOpsWF {
```

Comment:
This is a really nice example of the power of the composability of this abstraction, but it did leave me very confused for a while. For example, I wasn't exactly sure (and I'm still not exactly sure) why `using SymanzikOps = combineOps<computeStapleRectangleOps, KernelOps<StapOp, RectOp>>;` couldn't be `using SymanzikOps = combineOps<computeStapleRectangleOps, StapOp, RectOp>;`.

There may also be a typo where `using StapOp = ThreadLocalCache<Link, 0, computeStapleRectangeOps>;` should actually be `using StapOp = ThreadLocalCache<Link, 0, computeStapleOps>;` and it happens to work out because `computeStapleOps` and `computeStapleRectangeOps` are both `KernelOps<thread_array<int, 4>>;`... so maybe that would address my confusion.

In any case, I think overly-verbose documentation in this file and/or a wiki page that really digs into what's going on here as an example would be invaluable.

Reply:
Right now, combineOps only takes KernelOps or NoKernelOps types as template parameters, and combines them into a single combined KernelOps (or NoKernelOps) type as appropriate.  We could allow other types (like thread_array, etc.) as template arguments, but it adds to the complexity of the template programming, and makes it a bit harder to prevent users from accidentally using an unintended type (we would need to explicitly allow only the valid types, which I was trying to avoid so it is possible to add new ops types with only making minimal changes to the supporting infrastructure).  Yes, this does point out another place that needs better documentation, thanks.

As for using computeStapleOps vs. computeStapleRectangeOps, you caught me being a bit lazy there :) .  One branch uses computeStaple (Arg::wflow_type == QUDA_GAUGE_SMEAR_WILSON_FLOW) and the other computeStapleRectangle (Arg::wflow_type == QUDA_GAUGE_SMEAR_SYMANZIK_FLOW), so it really should choose between the two.  I'll fix that.

Reply:
The reasoning makes sense, thanks for explaining!

Reply:
Actually, the `StapOp` alias is only used in the case that `computeStapleRectangle` is called, so I think the code is correct as is (`StapOp` should use the offset of `computeStapleRectangeOps`).